docs/langref/relay_type.rst

```diff
@@ -0,0 +1,213 @@
+===================
+Relay's Type System
+===================
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have not yet described the type system. Relay is
+a statically typed and type-inferred language, allowing programs to
+be typed with a minimal requirement of explicit type information.
+
+Static types are useful when performing compiler optimizations because they
+communicate properties about the data we manipulate, such as runtime shape,
+data layout, and storage without needing to run the program. Additionally,
+static typing is useful for determining the 
+
+Relay's type system features a form of depending typing for shapes to
+replace the dynamic shape inference used in NNVM and most other machine
+learning IRs. Treating tensor shapes as types allows Relay to perform
+more powerful reasoning at compile time; in particular, Relay can statically
+reason about operations whose output shapes vary based on the input shapes
+in complex ways. Statically reasoning about shapes in this manner allows
+Relay to be ahead-of-time compiled and provides much more information about
+tensors for optimizations furhter in the compilation pipeline.
+
+Reasoning about tensor types in Relay is encoded using type relations, which means
+that the bulk of type checking in Relay is constraint solving (ensuring that all
+type relations are satisfied at call sites). Type relations offer a flexible and
+relatively simple way of making the power of dependent typing available in Relay
+without greatly increasing the complexity of Relay's type AST.
+
+Types
+=====
+
+Below we detail the language of types in Relay and how they are assigned to Relay expressions.
+
+Type
+~~~~
+
+The base type for all Relay types. All Relay types are sub-classes of this base type.
+
+See :py:class:`~tvm.relay.ty.Type` for its definition and documentation.
+
+Tensor Type
```

Comment:
Does zero-dim tensor, i.e. scalar, fall into this category? If so, could you please elaborate how it is represented in Relay? I think this is a very important improvement since it's not resolved in NNVM.

Reply:
Yes, scalars are also given tensor types. I will add a sentence making that explicit.